File: RootTools2/src/main/java/com/rarnu/tools/neo/xposed/MIUIRoot25.java
package com.rarnu.tools.neo.xposed;

import android.os.Bundle;
import android.os.Message;
import android.widget.Button;
import android.widget.TextView;
import de.robv.android.xposed.*;
import de.robv.android.xposed.callbacks.XC_InitPackageResources;
import de.robv.android.xposed.callbacks.XC_LayoutInflated;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

public class MIUIRoot25 implements IXposedHookInitPackageResources, IXposedHookLoadPackage {

    public static TextView warningText;
    public static Button accept;

    @Override
    public void handleLoadPackage(XC_LoadPackage.LoadPackageParam loadPackageParam) throws Throwable {
        XSharedPreferences prefs = new XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF);
        prefs.makeWorldReadable();
        prefs.reload();
        if (!prefs.getBoolean(XpStatus.KEY_ROOTCRACK, false)) {
            return;
        }
        if (!loadPackageParam.packageName.equals("com.miui.securitycenter")) {
            return;
        }
        XpUtils.findAndHookMethod("com.miui.permcenter.root.RootApplyActivity", loadPackageParam.classLoader, "onCreate", Bundle.class, new XC_MethodHook() {
            protected void afterHookedMethod(MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                if (accept == null) {
                    return;
                }
                int i = 0;
                while (i < 5) {
                    accept.performClick();
                    i += 1;
                }
            }
        });
        XpUtils.findAndHookMethod("com.miui.permcenter.root.c", loadPackageParam.classLoader, "handleMessage", Message.class, new XC_MethodReplacement() {
            protected Object replaceHookedMethod(MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                return null;
            }
        });
        XpUtils.findAndHookMethod("com.miui.permcenter.root.a", loadPackageParam.classLoader, "handleMessage", Message.class, new XC_MethodReplacement() {
            protected Object replaceHookedMethod(MethodHookParam paramAnonymousMethodHookParam) throws Throwable {
                return null;
            }
        });
    }


    @Override
    public void handleInitPackageResources(XC_InitPackageResources.InitPackageResourcesParam initPackageResourcesParam) throws Throwable {
        XSharedPreferences prefs = new XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF);
        prefs.makeWorldReadable();
        prefs.reload();
        if (!prefs.getBoolean(XpStatus.KEY_ROOTCRACK, false)) {
            return;
        }
        if (!initPackageResourcesParam.packageName.equals("com.miui.securitycenter")) {
            return;
        }
        initPackageResourcesParam.res.hookLayout("com.miui.securitycenter", "layout", "pm_activity_root_apply", new XC_LayoutInflated() {
            public void handleLayoutInflated(XC_LayoutInflated.LayoutInflatedParam paramAnonymousLayoutInflatedParam) throws Throwable {
                accept = (Button) paramAnonymousLayoutInflatedParam.view.findViewById(paramAnonymousLayoutInflatedParam.res.getIdentifier("accept", "id", "com.miui.securitycenter"));
                warningText = (TextView) paramAnonymousLayoutInflatedParam.view.findViewById(paramAnonymousLayoutInflatedParam.res.getIdentifier("warning_info", "id", "com.miui.securitycenter"));
                if (warningText != null) {
                    warningText.setLines(6);
                }
            }
        });
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/xposed/MIUIRoot25.kt
package com.rarnu.tools.neo.xposed

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Message
import android.widget.Button
import android.widget.TextView
import com.rarnu.xfunc.*
import de.robv.android.xposed.*

import de.robv.android.xposed.callbacks.XC_LayoutInflated


@SuppressLint("StaticFieldLeak")
private var warningText: TextView? = null
@SuppressLint("StaticFieldLeak")
private var accept: Button? = null

class MIUIRoot25Resources: XposedResource() {

    override fun hook(res: XposedRes) {
        val prefs = XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF)
        prefs.makeWorldReadable()
        prefs.reload()
        if (res.packageName == "com.miui.securitycenter") {
            if (prefs.getBoolean(XpStatus.KEY_ROOTCRACK, false)) {
                res.res.hookLayout("com.miui.securitycenter", "layout", "pm_activity_root_apply", object : XC_LayoutInflated() {
                    override fun handleLayoutInflated(paramAnonymousLayoutInflatedParam: LayoutInflatedParam) {
                        accept = paramAnonymousLayoutInflatedParam.view.findViewById(paramAnonymousLayoutInflatedParam.res.getIdentifier("accept", "id", "com.miui.securitycenter"))
                        warningText = paramAnonymousLayoutInflatedParam.view.findViewById(paramAnonymousLayoutInflatedParam.res.getIdentifier("warning_info", "id", "com.miui.securitycenter"))
                        warningText?.setLines(6)
                    }
                })
            }
        }
    }
}

class MIUIRoot25Pakage : XposedPackage() {
    override fun hook(pkg: XposedPkg) {
        val prefs = XSharedPreferences(XpStatus.PKGNAME, XpStatus.PREF)
        prefs.makeWorldReadable()
        prefs.reload()

        if (pkg.packageName == "com.miui.securitycenter") {
            if (prefs.getBoolean(XpStatus.KEY_ROOTCRACK, false)) {
                pkg.findClass("com.miui.permcenter.root.RootApplyActivity").findMethod("onCreate", Bundle::class.java).hook {
                    after {
                        result = null
                        if (accept == null) {
                            return@after
                        }
                        var i = 0
                        while (i < 5) {
                            accept?.performClick()
                            i += 1
                        }
                    }
                }
                pkg.findClass("com.miui.permcenter.root.c").findMethod("handleMessage", Message::class.java).hook { replace { result = null } }
                pkg.findClass("com.miui.permcenter.root.a").findMethod("handleMessage", Message::class.java).hook { replace { result = null } }
            }
        }
    }
}
